### ISSUE TITLE
Carthage update --use-submodules fails opaquely, needs better diagnostic messages

### ISSUE THREAD

**Issue by chrisco314** [Carthage/Carthage#792]
I am using carthage as an aid for maintaining internal dependencies.

I have in numerous contexts gotten the dreaded: `Parse error: expected submodule commit SHA in ls-tree output` error.  Previously, the problem was dangling references in .gitmodules.

This time around, I checked the above, tried nuking the shared carthage cache to no avail.  Ultimately, I determined that I had inadvertently checked in the Carthage build directory in my private framework as part of a batch check-in.

I am not making the case that the above should succeed, but given the number of error reports with the above error (and without knowing the challenges involved), I advocate - for the love of free time, for more specific error messages so a problem like this can be disambiguated from the other causes.

To reproduce, create new framework project, add some random Carthage dependency - like Alamofire, run carthage update, commit the whole thing (my original error), then try to pull said framework again using carthage.  The result is the above error with nary a clue as to the actual cause of the problem.  Adding more descriptive messages would help not just this error, but help disambiguate all the others.

Thanks!

**Comment by chrisco314** [comment]
Clearly I don't have a complete grasp of git and submodules.  I was having a difficult time teasing apart problems in my framework vs client project.  I thought I had solved it with the actions stated in the initial post - but it worked on one machine, but on the other, I was still getting the same message.  

I ended up nuking the framework's .gitmodules file, re-running 

`carthage update --use-submodules --no-build`, 

committing the updated `.gitmodules` file along with the directory structure, which had the net effect of adding one dependency.

Then, in the client repo, after repeatedly getting the `...SHA in ls-tree` error, I manually removed the submodules from instrutions provided here: https://gist.github.com/kyleturner/1563153

At which point I again ran ``carthage update --use-submodules --no-build`, and again committed the newly created submodules, and then committed that.  At that point, all was well with carthage and git.

**Comment by mdiep** [comment]
Glad you figured it out! :smile:

**Comment by chrisco314** [comment]
@mdiep Is there anyway to establish more of a context though in the error message?  as in "\(error) wen trying to \(context)"?

**Comment by mdiep** [comment]
Possibly. I'm not sure.

**Comment by chrisco314** [comment]
So this happened to me again.  It appeared to be introduced by my use of Atlassian's SourceTree that I was using to browse all the changes.  

SourceTree showed appeared to think that there was a new submodule / folder, and that was in the index, and had the same submodule folder as deleted in the non staged section.  Not wanting to mess with things, I deselected the add-submodule (folder) from the index, and it seemed to combine / confuse the two entries.  When I immediately tried to restore via carthage, I got the above error.
